packages/console/src/hooks/use-user-default-tenant-id.ts

```diff
@@ -17,32 +17,40 @@ const key = 'defaultTenantId';
 const useUserDefaultTenantId = () => {
   const { data, update: updateMeCustomData } = useMeCustomData();
   const { tenants, currentTenantId } = useContext(TenantsContext);
+  const storedId = useMemo(
+    () => trySafe(() => z.object({ [key]: z.string() }).parse(data)[key]),
+    [data]
+  );
+  // Ensure update the same tenant ID to the stored ID only once.
+  const [updatedTenantId, setUpdatedTenantId] = useState('');
 
   const defaultTenantId = useMemo(() => {
-    const storedId = trySafe(() => z.object({ [key]: z.string() }).parse(data)[key]);
-
     // Ensure the stored ID is still available to the user.
     if (storedId && tenants.some(({ id }) => id === storedId)) {
       return storedId;
     }
 
     // Fall back to the first tenant ID.
     return tenants[0]?.id;
-  }, [data, tenants]);
+  }, [storedId, tenants]);
 
   const updateIfNeeded = useCallback(async () => {
-    if (currentTenantId !== defaultTenantId) {
+    if (currentTenantId !== storedId && currentTenantId !== updatedTenantId) {
```

Comment:
Feels like there are too many tenant IDs to manage: currentTenantId, storedId, updatedTenantId. I'm already a little bit confused

Reply:
i can add some comments here. I've tried my best to reduce variables

Reply:
Maybe we can write a `useEffect` to update `updatedTenantId` whenever the custom data changes, so that we can get rid of `storedId`, and only need to check if `currentTenantId` equals `updatedTenantId`?

Reply:
let me double check

Reply:
we still need a `storedId` since it should be initial value for `updatedTenantId `

Reply:
`useEffect` still serves this purpose, I think

Reply:
yes, but then it has nearly no difference between the implementations? only saves an `if`. becomes more clear though.

Reply:
please check if the update is expected